# The pandas library

With Python gaining popularity in data science, many  libraries have been created and maintained by different developer teams. These are collectively called **third party libraries**.

If the Python environment you are using do not already include a third party library you want to use, you can use a **pip install command** to install the library.

The ! before `pip` in the following block is to let Google Colab know that this is not a Python statement but a system command.

In [ ]:
!pip install library_name_of_your_choice
#Do not run this block since the above is provided as a template for installing a new third-party library of your choice.

The [pandas library](https://pandas.pydata.org/) provides a convenient way to analyze and clean data. It is one of the common third-party libraries to work with tabular data. The term **pandas** is derived from the term "**panel data**", referring to data sets that include observations over multiple time periods for the same individuals.  Fortunately, Google Colab already includes this library in its environment. If you are running Python in an environment that does not already support this library, you will need to run the following to install it.

In [ ]:
!pip install pandas
#Do not run this block since Google Colab already supports the pandas library

To use the functions from the `pandas` library in our program, it is important and common to include the following statement so that we can use `pd` as the abbreviation of `pandas` when calling functions from the `pandas` library.

In [1]:
import pandas as pd

## DataFrame and Series

Social scientists often use surveys to collect data for their research questions. Data collected from a survey are often stored in a **tabular** format where
*   Each **column** corresponds to a question on the survey and contains responses to the question from all respondants.
*   Each **row** corresponds to one respondant and covers responses for all questions from the respondant.

Similarly, a crime reporting data set is likely to be in a **tabular** format where
*   Each **column** corresponds to one aspect of a crime and contains details for that aspect from all crimes reported.
*   Each **row** corresponds to one crime and covers details for all aspects of the crime.

`DataFrame` and `Series` are two popular modules from the `pandas` library. They share a lot of functions.
*  We will be using a `DataFrame` objects to hold one **tabular** dataset.
*  One **row** or one **column** of a `DataFrame` would be considered as a `Series` object.

# Create a DataFrame Object

The following code block illustrates how to create a `DataFrame` object. Run the code to see how the rows and columns of the `DataFrame` object are organized.

The `population_datat` variable uses a **dictionary** object to set up the data and their structure for a `DataFrame` object. A dictionary uses a pair of curly brackets to enclose a list of **key:value** pairs.
*   The **key** before each colon specifies the name of a column.
*   The **value** after each colon specifies the **list** of values to be stored under the related column heading. The number of elements in the list dictate the number of rows for the table. When creating a dataset for a `DataFrame` object, it is important that **all lists share the same number of elements**.

Line 3 demonstrates how a new `DataFrame` object can be created using the `pd.DataFrame` function. It takes a dictionary object (`population_dataset`) as its argument. The assignment statement store the new `DataFrame` object into the variable `population_df`.

In [2]:
#source of data: https://www.pewresearch.org/social-trends/2014/04/10/next-america/#the-generational-divide
import pandas as pd
population_data = {
    'Year':[1960, 1970, 1980, 1990, 2000, 2010, 2020, 2030, 2040, 2050, 2060],
    'White':[.85, .83, .80, .76, .69, .64, .60, .55, .51, .47, .43],
    'Black': [.10, .11, .12, .12, .12, .12, .13, .13, .13, .13, .13],
    'Hispanic': [.04, .05, .06, .09, .13, .16, .19, .22, .25, .28, .31],
    'Asian': [.01, .01, .02, .03, .04, .05, .06, .06, .07, .08, .08],
    'Others': [.00, .00, .01, .01, .02, .03, .03, .04, .04, .05, .06]
    }
population_df = pd.DataFrame(population_data)
population_df

#make sure that you run this code block before running the code blocks below

,Year,White,Black,Hispanic,Asian,Others
0,1960,0.85,0.10,0.04,0.01,0.00
1,1970,0.83,0.11,0.05,0.01,0.00
2,1980,0.80,0.12,0.06,0.02,0.01
3,1990,0.76,0.12,0.09,0.03,0.01
4,2000,0.69,0.12,0.13,0.04,0.02
5,2010,0.64,0.12,0.16,0.05,0.03
6,2020,0.60,0.13,0.19,0.06,0.03
7,2030,0.55,0.13,0.22,0.06,0.04
8,2040,0.51,0.13,0.25,0.07,0.04
9,2050,0.47,0.13,0.28,0.08,0.05


We can also use the print function to show the content of a DataFrame object.

In [ ]:
print(population_df)

    Year  White  Black  Hispanic  Asian  Others
0   1960   0.85   0.10      0.04   0.01    0.00
1   1970   0.83   0.11      0.05   0.01    0.00
2   1980   0.80   0.12      0.06   0.02    0.01
3   1990   0.76   0.12      0.09   0.03    0.01
4   2000   0.69   0.12      0.13   0.04    0.02
5   2010   0.64   0.12      0.16   0.05    0.03
6   2020   0.60   0.13      0.19   0.06    0.03
7   2030   0.55   0.13      0.22   0.06    0.04
8   2040   0.51   0.13      0.25   0.07    0.04
9   2050   0.47   0.13      0.28   0.08    0.05
10  2060   0.43   0.13      0.31   0.08    0.06


Here is another example for creating a new DataFrame object.
*  'category', 'planned', and 'actual' are the names for the columns.
*  A colon is used to separate a column name and the list of values for that column.
*  The items of the lists match by their positions in the list. They make up the respective rows of the table. For example, the first row of the table has 'Rent' for the 'category' column, 950 for the 'planned' column, and 950 for the 'actual' column.
*  It is important to make sure all lists contain the same number of values because they make up the rows of the table.

In [3]:
budget_data = {
    "category": ["Rent", "Groceries", "Dining Out", "Phone", "Transport", "Misc"],
    "planned": [1500, 220, 120, 40, 60, 80],
    "actual":  [1300, 215, 138, 40, 68, 57]
}
budget_df = pd.DataFrame(budget_data)
budget_df

,category,planned,actual
0,Rent,1500,1300
1,Groceries,220,215
2,Dining Out,120,138
3,Phone,40,40
4,Transport,60,68
5,Misc,80,57


## Practice

In [ ]:
# Use data from HDPulse (https://hdpulse.nimhd.nih.gov/data-portal/home) to create a small data frame using data from
# five California counties of your choice
# The first column should be the names of the counties.
# The second and third columns should be related data for these counties



# Initial inspections

Before performing any data analysis or cleaning, it's crucial to perform an initial inspection of your pandas DataFrame. This helps you understand its structure, content, data types, and identify potential issues such as missing values or incorrect data formats.

In the following subsections, we will introduce and demonstrate the use of several useful pandas functions for this initial inspection.

## The head and tail function

The `head()` function is used to display the first few rows of a DataFrame. By default, it shows the first 5 rows. You can specify the number of rows to display by passing an integer argument to the function. This is useful for getting a quick look at the structure and content of your data.

In [ ]:
population_df.head()

,Year,White,Black,Hispanic,Asian,Others
0,1960,0.85,0.10,0.04,0.01,0.00
1,1970,0.83,0.11,0.05,0.01,0.00
2,1980,0.80,0.12,0.06,0.02,0.01
3,1990,0.76,0.12,0.09,0.03,0.01
4,2000,0.69,0.12,0.13,0.04,0.02


In [ ]:
population_df.head(3)

,Year,White,Black,Hispanic,Asian,Others
0,1960,0.85,0.10,0.04,0.01,0.00
1,1970,0.83,0.11,0.05,0.01,0.00
2,1980,0.80,0.12,0.06,0.02,0.01


We can save the result of the head() function into a new DataFrame object.

In [ ]:
first_3_years = population_df.head(3)
print(first_3_years)

   Year  White  Black  Hispanic  Asian  Others
0  1960   0.85   0.10      0.04   0.01    0.00
1  1970   0.83   0.11      0.05   0.01    0.00
2  1980   0.80   0.12      0.06   0.02    0.01


## The tail() function

The `tail()` function is similar to `head()`, but it displays the *last* few rows of a DataFrame. By default, it also shows the last 5 rows. You can specify the number of rows to display by passing an integer argument to the function. This is useful for checking the end of your dataset and can sometimes reveal patterns or issues at the end of the data collection.

In [ ]:
population_df.tail()

,Year,White,Black,Hispanic,Asian,Others
6,2020,0.60,0.13,0.19,0.06,0.03
7,2030,0.55,0.13,0.22,0.06,0.04
8,2040,0.51,0.13,0.25,0.07,0.04
9,2050,0.47,0.13,0.28,0.08,0.05
10,2060,0.43,0.13,0.31,0.08,0.06


Demonstrate how to use the `tail` function to display a specific number of rows of the `population_df` DataFrame.


In [ ]:
population_df.tail(3)

,Year,White,Black,Hispanic,Asian,Others
8,2040,0.51,0.13,0.25,0.07,0.04
9,2050,0.47,0.13,0.28,0.08,0.05
10,2060,0.43,0.13,0.31,0.08,0.06


We can save the result of the tail function into a new DataFrame object.

In [ ]:
last_3_years = population_df.tail(3)
print(last_3_years)

    Year  White  Black  Hispanic  Asian  Others
8   2040   0.51   0.13      0.25   0.07    0.04
9   2050   0.47   0.13      0.28   0.08    0.05
10  2060   0.43   0.13      0.31   0.08    0.06


### Practice

In [ ]:
# Call the head function to show the first two rows of your county data


In [ ]:
# Call the tail function to show the last two rows of your county data


## The shape attribute

The `shape` attribute returns a tuple representing the dimensions of the DataFrame, in the format (number of rows, number of columns). This gives you a quick understanding of the size of your dataset.

In [ ]:
population_df.shape

(11, 6)

The above result shows that there are 11 rows and 6 columns in population_df.

In [ ]:
budget_df.shape

(6, 3)

The above result shows that there are 6 rows and 3 columns in budget_df.

## The columns attribute

The `columns` attribute returns the column labels of the DataFrame. This is useful for quickly seeing the names of all columns in your dataset.

In [ ]:
population_df.columns

Index(['Year', 'White', 'Black', 'Hispanic', 'Asian', 'Others'], dtype='object')

In [ ]:
budget_df.columns

Index(['category', 'planned', 'actual'], dtype='object')

## The info() function

The `info()` function provides a concise summary of a DataFrame. This includes:
*   The index dtype and columns.
*   Non-null values and their counts for each column.
*   Data types (dtypes) for each column.
*   Memory usage of the DataFrame.

This function is essential for a quick overview of your data's structure, helping you identify missing values (columns with fewer non-null entries than the total number of rows) and check if the data types are as expected.

In [ ]:
population_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11 entries, 0 to 10
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Year      11 non-null     int64  
 1   White     11 non-null     float64
 2   Black     11 non-null     float64
 3   Hispanic  11 non-null     float64
 4   Asian     11 non-null     float64
 5   Others    11 non-null     float64
dtypes: float64(5), int64(1)
memory usage: 660.0 bytes


Here are a few things we can learn about the dataset from the above result of the info function.

*  "RangeIndex: 11 entries, 0 to 10": This means there are 11 rows in the dataset. The index for the first row is 0. The index for the last row is 10.
*  "Data columns (total 6 columns)": This means there are 6 columns in the data set.
*  " 0   Year      11 non-null     int64": This means that the first column's name is 'Year'. There are 11 non-null values in this column. The data type for the values in the column is int64, i.e. integer.
*  "1   White     11 non-null     float64": This means that the second column's name is 'White'. There are 11 non-null values in this column. The data type for the values in the column is float64, i.e. floating point values.
*  "memory usage: 660.0 bytes": This indicates that the dataset takes up 660 bytes of memory.

In [ ]:
budget_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6 entries, 0 to 5
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   category  6 non-null      object
 1   planned   6 non-null      int64 
 2   actual    6 non-null      int64 
dtypes: int64(2), object(1)
memory usage: 276.0+ bytes


Here are a few things we can say about the budget dataframe.

*  There are 6 entries (rows) in the table. The first row has 0 as its index and the last row has 5 as its index.
*  There are three columns. The first column is 'category'. It has 6 non-null values. The data type for the values in this column is object, indicating string/text. The second column is 'planned'. The data type for the values in this column is int64, i.e. integer.

### Practice

In [ ]:
# Call the shape attribute to show the number of rows and columns of your county dataframe.


In [ ]:
# Call the columns attribute to show the column names of your county dataframe


In [ ]:
# Call the info function to show a summary of your county dataframe


**Add a text block** to explain what you learn about your county dataframe object.
*  How many rows and columns are there in the dataset?
*  What are the names of the columns?
*  What are the data types of the values in respective columns?

# The describe() function

The `describe()` function generates descriptive statistics of the DataFrame's numerical columns. This includes count, mean, standard deviation, minimum, maximum, and quartile values. This function is very useful for getting a statistical summary of your data and identifying potential outliers or distributions.

In [ ]:
population_df.describe()

,Year,White,Black,Hispanic,Asian,Others
count,11.000000,11.000000,11.000000,11.000000,11.000000,11.000000
mean,2010.000000,0.648182,0.121818,0.161818,0.046364,0.026364
std,33.166248,0.148850,0.009816,0.095793,0.026181,0.020136
min,1960.000000,0.430000,0.100000,0.040000,0.010000,0.000000
25%,1985.000000,0.530000,0.120000,0.075000,0.025000,0.010000
50%,2010.000000,0.640000,0.120000,0.160000,0.050000,0.030000
75%,2035.000000,0.780000,0.130000,0.235000,0.065000,0.040000
max,2060.000000,0.850000,0.130000,0.310000,0.080000,0.060000


There are too many digits after the decimal points, making the results hard to understand. We can use pd.options.display.float_format = "{:.2f}".format to help specify the number of digits after the decimal place.

In [4]:
pd.options.display.float_format = "{:.2f}".format
population_df.describe()

,Year,White,Black,Hispanic,Asian,Others
count,11.00,11.00,11.00,11.00,11.00,11.00
mean,2010.00,0.65,0.12,0.16,0.05,0.03
std,33.17,0.15,0.01,0.10,0.03,0.02
min,1960.00,0.43,0.10,0.04,0.01,0.00
25%,1985.00,0.53,0.12,0.07,0.03,0.01
50%,2010.00,0.64,0.12,0.16,0.05,0.03
75%,2035.00,0.78,0.13,0.23,0.07,0.04
max,2060.00,0.85,0.13,0.31,0.08,0.06


Here are some observations we can make based on the above descriptive statistics:

*  The **count** row shows that there are 11 rows in the table.
*  The **min** row shows the smallest value for each column. The earliest data is for 1960. The lowest percentage for white is 0.43, i.e. 43% while the lowest percentage for Hispanic is 0.04, i.e. 4%. It is important to note that we do not know from the above descriptive statistics which year these percentages belong to.
*  The **50%** row shows the median value for each column. For example, 50% of the time, White accounts for at most 64% of the population.
*  The **max** row shows the largest value for each column. The latest data is for 2060, as a prediction. The highest percentage for Black is 13%, while the highest percentage for Asian is 8%. Again, we do not know which year these percentags belong to by simply looking at these statistics.
*  The **mean** row shows the average value for each column. On average, White has higher percentage than other ethnicities.
*  The **std** row shows the standard deviation for each column. The small std value for Black indicates not much change in their percentage. The higher std value for White indicates more change in their percentage.

## Practice

In [ ]:
# Call the describe function to show descriptive statistics of your county data


**Add a text block** to explain what you learn about your county data based on the stats.

*  What are the average and median values for respective columns?
*  What are the smallest, largest values for respective columns?
*  What are the ranges and standard deviations for respective columns?